File: patterns/state-machine-driven-groups/rules-python/state_machine_driven_groups_complex.py
from openhab import rule, Registry
from openhab.triggers import ItemStateChangeTrigger


@rule(triggers=[ItemStateChangeTrigger("vTimeOfDay")])
class SetLightsBasedOnTimeOfDay:
    def execute(self, module, input):
        time_of_day = str(Registry.getItem("vTimeOfDay").getState())

        for light in Registry.getItem("gLights").getAllMembers():
            setting = next(
                (s for s in Registry.getItem("gSettings").getAllMembers()
                 if s.getName() == light.getName() + "_" + time_of_day),
                None,
            )
            if setting is not None:
                setting_state = setting.getState()
                if str(setting_state) in ("NULL", "UNDEF"):
                    self.logger.warn(light.getName() + ": Sollwert fuer " + time_of_day + " noch nicht gesetzt")
                    continue
                light.sendCommand(str(setting_state))
                self.logger.info(light.getName() + " -> " + str(setting_state))

File: patterns/state-machine-driven-groups/rules-js/state-machine-driven-groups-complex.js
rules.JSRule({
  name: "Set lights based on Time of Day",
  triggers: [triggers.ItemStateChangeTrigger("vTimeOfDay")],
  execute: (event) => {
    const timeOfDay = items.getItem("vTimeOfDay").state;

    items.getItem("gLights").members.forEach((light) => {
      const setting = items.getItem("gSettings").members
        .find((s) => s.name === light.name + "_" + timeOfDay);
      if (!setting) {
        // Keine Einstellung fuer diese Tageszeit vorgesehen - kein Fehler
      } else if (setting.state === "NULL" || setting.state === "UNDEF") {
        console.warn(light.name + ": Sollwert fuer " + timeOfDay + " noch nicht gesetzt");
      } else {
        light.sendCommand(String(setting.state));
        console.log(light.name + " -> " + setting.state);
      }
    });
  }
});
